### DISCUSSION TITLE
How are plugins configured when using charon-systemd?

### DISCUSSION THREAD

**Discussion by Jafaral** [strongswan/strongswan#1751] [Q&A]
I'm  testing version 6.0, and I built and configured strongswan following the instructions here:

https://docs.strongswan.org/docs/6.0/daemons/charon-systemd.html

Now that I don't have `charon`, but instead I have `charon-systemd`, how does that affect plugin configuration?

strongswan.conf has a charon section, and now it  also has a charon-systemd section. There are also modular plugin config files  under `/etc/strongswan.d/charon/`.
  - Are those modular files apply to charon-systemd under the same old charon path? 
  -  what If I configure a plugin in strongswan.conf under `charon-systemd` section, does that override config under  `/etc/strongswan.d/charon/`?

I configured openssl to test as follows:
```
cat /etc/strongswan.conf
charon-systemd {
   plugins {
      openssl {
         load = yes
         fips_mode = 1
      }
   }
}
```

which doesn't seem to be loaded if I just fo `swanctl --reload-settings`, but it does seem to work when I restart the service.

The page https://docs.strongswan.org/docs/6.0/plugins/plugins.html doesn't cover the topic of having multiple configs in two places, nor it covers the relation with `charon-systemd` so I'm a bit at a lost here  trying to play a guessing game.

Thank you in advance.

**Comment by tobiasbrunner** [comment]
> * Are those modular files apply to charon-systemd under the same old charon path?

Yes, _charon-systemd_ inherits all `charon.*` settings from strongswan.conf.

> * what If I configure a plugin in strongswan.conf under `charon-systemd` section, does that override config under  `/etc/strongswan.d/charon/`?

Yes.

> which doesn't seem to be loaded if I just fo `swanctl --reload-settings`, but it does seem to work when I restart the service.

Plugins are only loaded when the daemons are (re-)started.

> The page https://docs.strongswan.org/docs/6.0/plugins/plugins.html doesn't cover the topic of having multiple configs in two places,

What do you mean with "multiple configs in two places"?

**Comment by Jafaral** [comment]
@tobiasbrunner , I meant that I didn't see a discussion of a plugin configuration in strongswan.conf vs a separate file under the charon directory, but maybe I just missed it. also, the new topic configs of charon vs charon-systemd. 

Also, the swanctl [doc](https://docs.strongswan.org/docs/5.9/swanctl/swanctlReloadSettings.html) states that `swanctl --reload-settings` does reload strongswan.conf, but according to your response above, that is not the case for all configs. Are there any configurations beside plugins that require restarting to apply?

**Reply by tobiasbrunner** [reply]
> I meant that I didn't see a discussion of a plugin configuration in strongswan.conf vs a separate file under the charon directory, but maybe I just missed it. 

That's doesn't really matter but obviously depends on includes and refs how the final config will look like.

> Are there any configurations beside plugins that require restarting to apply?

Yes, lots of them.

**Comment by Jafaral** [comment]
Some  of that is actually covered in strongswan.conf [doc](https://docs.strongswan.org/docs/6.0/config/strongswanConf.html#_reloading). 

Is there an easy way to tell what configs need restart vs reload? I like to use reload whenever possible. 

> Yes, lots of them.
any chance those are documented? I like to reload vs restart when possible. 



> Yes, _charon-systemd_ inherits all `charon.*` settings from strongswan.conf.
> 
> > * what If I configure a plugin in strongswan.conf under `charon-systemd` section, does that override config under  `/etc/strongswan.d/charon/`?
> 
> Yes.
> 
> >

The documentation [here](https://docs.strongswan.org/docs/6.0/config/strongswanConf.html#_including_files) suggests that loaded files extend/replace existing configs which seems to be not the case for plugins according to your answer and also my testing, ie, with plugins, if a plugin is configured  in strongswan.conf, the plugin separate file is ignored completely. Is that correct?

**Reply by tobiasbrunner** [reply]
> Is there an easy way to tell what configs need restart vs reload?

Other than reading the source code? No.

>  if a plugin is configured in strongswan.conf, the plugin separate file is ignored completely

That depends entirely on the rest of strongswan.conf. The config snippets under [`strongswan.d`](https://docs.strongswan.org/docs/5.9/config/strongswanDir.html) are only used if there are `include` statements in strongswan.conf that reference them.

**Comment by Jafaral** [comment]
> > Is there an easy way to tell what configs need restart vs reload?
> 
> Other than reading the source code? No.

bummer!  :-)


> > if a plugin is configured in strongswan.conf, the plugin separate file is ignored completely
> 
> That depends entirely on the rest of strongswan.conf. The config snippets under [`strongswan.d`](https://docs.strongswan.org/docs/5.9/config/strongswanDir.html) are only used if there are `include` statements in strongswan.conf that reference them.

I see, so a separate plugin config file would still override strongswan.conf if strongswan does include that separate file. 

OFF-TOPIC: I'm testing this with 6.0, the latest tag is beta3, but I do see new updates made it to the six-beta branch, any ETA for beta4 or an official 6.0 release?

**Reply by tobiasbrunner** [reply]
> I see, so a separate plugin config file would still override strongswan.conf if strongswan does include that separate file.

Only if strongswan.conf makes it so (also depends on the order of the settings/includes in the file). Note that strongSwan's executables don't include anything but strongswan.conf (they know nothing about individual plugin files etc.).

> any ETA for beta4 or an official 6.0 release?

No, currently not.
